WTG.Analyzers/Analyzers/Array/ArrayAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using WTG.Analyzers.Utils;
+
+namespace WTG.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public sealed class ArrayAnalyzer : DiagnosticAnalyzer
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rules.PreferArrayEmptyOverNewArrayConstructionRule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(CompilationStart);
+		}
+
+		static void CompilationStart(CompilationStartAnalysisContext context)
+		{
+			var cache = new FileDetailCache();
+
+			context.RegisterSyntaxNodeAction(
+				c => Analyze(c, cache),
+				SyntaxKind.ArrayCreationExpression);
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
+			{
+				return;
+			}
+
+			var syntax = (ArrayCreationExpressionSyntax)context.Node;
+
+			if (syntax.FirstAncestorOrSelf<AttributeArgumentSyntax>() != null)
+			{
+				// Ignore empty arrays in attributes. Array.Empty<T>() cannot be used here.
+				return;
+			}
+
+			var ranks = syntax.Type.RankSpecifiers;
+			if (ranks.Count == 0)
+			{
+				// Incomplete code?
+				return;
+			}
+
+			var rank = ranks[0];
+			if (rank.Sizes.Count != 1)
+			{
+				// Ignore multi-dimensional top-level arrays.
+				return;
+			}
+
+			foreach (var size in rank.Sizes)
+			{
+				switch (size.Kind())
+				{
+					case SyntaxKind.OmittedArraySizeExpression:
+						if (syntax.Initializer == null)
+						{
+							// If there's an initializer, we check with the initializer expression count below.
+							// If there's no initializer, this is a compiler error anyhow, so don't bother the
+							// developer with an analyzer result.
+							return;
+						}
+						break;
+
+					case SyntaxKind.CharacterLiteralExpression:
+					case SyntaxKind.NumericLiteralExpression:
+					case SyntaxKind.CastExpression:
+					case SyntaxKind.UnaryMinusExpression:
+					case SyntaxKind.UnaryPlusExpression:
+						var constant = context.SemanticModel.GetConstantValue(size, context.CancellationToken);
+						if (constant.HasValue && !IsZeroLiteral(constant.Value))
+						{
+							return;
+						}
+						break;
+
+					default:
+						return;
+				}
+			}
+
+			if (syntax.Initializer?.Expressions.Count > 0)
+			{
+				return;
+			}
+
+			context.ReportDiagnostic(Rules.CreatePreferArrayEmptyOverNewArrayConstructionDiagnostic(context.Node.GetLocation()));
+		}
+
+		static bool IsZeroLiteral(object value)
+		{
+			switch (value)
```

Comment:
I would probably put the more common cases (eg. int) first and the crazy cases (eg. char) last.

int > uint > long > ulong > short > ushort > byte > sbyte > char

Reply:
Current order is integers small to large (signed first), then char. I personally find it reads better - though I'd also be open to putting char in with byte/sbyte since it's traditionally a byte (yes I know in .NET it's two bytes).

Reply:
My thinking is that in the vast majority of cases, value will be an int, so we should check for that first. I also put the rest in order of what I think the frequency would be, but I think the non-int literal case will be so uncommon that it doesn't really matter as long as int comes first :)

Reply:
Oh that's right, this kind of switch isn't a jump table, so order will affect performance. 😅